pruner/state.go

```diff
@@ -0,0 +1,81 @@
+package pruner
+
+import (
+	"fmt"
+	"sync"
+
+	quamina "quamina/lib"
+)
+
+// State represents the required capabilities for maintaining the set
+// of live patterns.
+type State interface {
+	Add(x quamina.X, pattern string) error
+	Del(x quamina.X) (bool, error)
+
+	// Iterate calls the given function for every stored pattern.
+	Iterate(func(x quamina.X, pattern string) error) error
+
+	// Get returns the pattern for the given X.
+	//
+	// Since a pattern can't be the empty string, that zero value
+	// indicates no corresponding pattern.
+	Get(x quamina.X) (string, error)
+}
+
+// MemState is a State that is just a map (with a RWMutex).
+type MemState struct {
+	lock sync.RWMutex
+	m    map[quamina.X]string
+}
+
+func NewMemState() *MemState {
+	// Accept initial size as a parameter?
+	return &MemState{
+		m: make(map[quamina.X]string),
+	}
+}
+
+var ErrExists = fmt.Errorf("pattern already exists for that X")
+
+func (s *MemState) Add(x quamina.X, pattern string) error {
+	s.lock.Lock()
+	var err error
+	if _, have := s.m[x]; have {
```

Comment:
Hmm, interesting semantics here. Quamina Matcher won't stop you adding multiple different Pattern strings with the same X value, in fact that's how you can get an OR effect.  Which, to be fair, the README doesn't say, oops sorry.  I think pruner should be consistent with Quamina here?

Reply:
Oh, wow. I should have checked that! Yeah, sure seems like we should be consistent here. Will fix.

Reply:
Fixed.